### PR TITLE
Use sparse - weights as arg and not kwarg.

### PR DESCRIPTION
**Pull request by aulemahal**
I have to regrid regional climate model data to the grid of ERA-Land over all of Canada. As one can guess, these are two very large grids and the weights, even in a sparse array, are of significant size. I have to compute this on a shared machine, where I am limited memory-wise more than CPU-wise. I found the following problem:

The regrid call use `xr.apply_ufunc` and the  weights are passed _as kwargs_ to `_regrid_array`. In the dask machinery, kwargs are copied and passed as-is. Thus, for each "regrid" tasks a full copy of the weights is made. In my case, weights are ~ 30 Mo and there are 345 chunks, yielding 10 Go only for that part. That's on the scheduler's memory, and not on the workers', so it just sits there until everything is done. You have to add the workers' RAM to this to get the total memory usage.

My idea was to pass the weights as an _argument_ of the `apply_ufunc` call, so that it is not copied on all tasks. Problem is that it needs to be an object xarray can deal with : cue in [`sparse`](https://sparse.pydata.org/en/stable/). 

This PR modifies the regridder object so that `Regridder.weights` is an xarray backed by a `sparse.COO` array, instead of directly being a scipy.sparse array. This allows to pass it a an arg and, potentially, decreases the memory usage a lot.

**Problem**: As it is now, dask feels the need to use smaller intermediate chunks because the sparse chunks looks so large. I'm guessing dask estimates the chunk footprint using the size and the dtype. To overcome that problem, I set the following:

```python3
from dask import config
config.set({'array.chunk-size': '500GiB'})
```

Without that line, the number of tasks increases a lot and the scheduler seems to have difficulties to even start the computation. With the line, everything works. It is slightly slower (slightly more tasks) then before, but the memory usage is a lot lower.

See:
`ds_in` is 300x340 (lat / lon) with 55114 timesteps divided in 345 chunks of 160. Each chunk is 125 MB.
`ds_out` is 415x884, with the same timesteps and chunking. Each chunk is 225 MB. (output is coerced to float32).

In the two plots below, the black line is the main process (including the dask scheduler) and the red line is the dask worker computing the regridding.

Here is the memory profiling with the master branch:
master

And here with this branch + the dask config change:
sparse_config

What do you say? Is there a way to circumvent the dask.config change?

I already fixed most of the tests, but this still needs some work. It would be nice to only use `sparse` and remove all calls to `scipy`.

### CODE REVIEW

**Review by huard — APPROVED** [review]
LGTM, but not seeing the big picture, so you most certainly will need more competent reviewers.

File: xesmf/frontend.py [hunk 1]
```diff
@@ -632,9 +634,11 @@ def to_netcdf(self, filename=None):
         """Save weights to disk as a netCDF file."""
         if filename is None:
             filename = self.filename
-        w = self.weights
+        w = self.weights.data
         dim = 'n_s'
-        ds = xr.Dataset({'S': (dim, w.data), 'col': (dim, w.col + 1), 'row': (dim, w.row + 1)})
+        ds = xr.Dataset(
+            {'S': (dim, w.data), 'col': (dim, w.coords[1, :] + 1), 'row': (dim, w.coords[0, :] + 1)}
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xesmf/frontend.py:640]: Yep. `self.weights` is a DataArray, `self.weights.data = w` is a  `sparce.COO` array and `w.data` is a 1D numpy array, of only the sparse values.

File: xesmf/smm.py [hunk 2]
```diff
@@ -173,11 +185,12 @@ def _combine_weight_multipoly(weights, indexes):
...
     columns = []
+    # The list.append ensures each summed column has the index of the value in `indexes`.
     for i in range(indexes.max() + 1):
-        js = np.where(indexes == i)[0]
-        wi = weights[..., js[0]]
-        for j in js[1:]:
-            wi = wi + weights[..., j]
-        columns.append(wi)
-    return sps.hstack(columns)
+        # Sum the colums with the same indexes in `indexes`
```
Comment (raphaeldussin) [inline xesmf/smm.py:192]: looks more readable, :+1:

**Review by raphaeldussin — CHANGES_REQUESTED** [review]
a few minor issues, but should be easy to fix

File: xesmf/smm.py [hunk 3]
```diff
@@ -6,19 +6,20 @@
...
 
 def read_weights(weights, n_in, n_out):
     """
-    Read regridding weights into a scipy sparse COO matrix.
+    Read regridding weights into a DataArray (sparse COO matrix).
 
     Parameters
     ----------
-    weights : str, Path, xr.Dataset
+    weights : str, Path, xr.Dataset, xr.DataArray, scipy.sparse.coo_matrix, sparse.COO
```
Comment (raphaeldussin) [inline xesmf/smm.py:19]: one last nitpick: `scipy.sparse.coo_matrix` do not seem supported by the code.
Do we want to add support with something like

```python
elif isinstance(weights, scipy.sparse.coo_matrix):
    # convert to dataarray
```

or remove from docstring?

**Review by raphaeldussin — CHANGES_REQUESTED** [review]

File: xesmf/smm.py (hunk 3 above)
Comment (aulemahal) [reply xesmf/smm.py:19]: oh woups! That isinstance cal was the only place in all xesmf that required the explicit dependency to  `scipy`, so I preferred to remove it. I'll remove it from the docstring.

**Review by raphaeldussin — APPROVED** [review]

**Comment by jhamman** [comment]
@pangeo-data/xesmf - This seems like a very useful feature. Any chance we can get a release of xESMF to get this out in the wild? Thanks!

**Comment by huard** [comment]
@jhamman Done

**Comment by jhamman** [comment]
Amazing! Thanks @huard!

**Pull request merged**